docs/protocol/core/v3.0.rst

```diff
@@ -95,7 +96,27 @@ provide.
 Node names
 ----------
 
-TODO define constraints on node names
+Except for the root node, each node in a hierarchy must have a name,
+which is a string of ASCII characters. To ensure consistent behaviour
+across different storage systems, the following constraints apply to
+node names:
+
+* must not be the empty string ("")
+
+* must consist only of characters in the sets `a-z`, `A-Z`, `0-9`,
+  `-_.`
```

Comment:
Is this too restrictive? I was thinking that node names are used in file paths when using file system storage, and so we should ensure no characters are used that could be problematic within a file system path. But are there other characters that people use in dataset naming that would be safe to include?